### PR TITLE
fix(ivy): TestBed not unwrapping imports array function when overriding provider

### LINKED ISSUE #34623: Ivy: unit test failed due to 'moduleDef.imports is not iterable'.
# 🐞 bug report

### Affected Package

The issue is caused by package @angular/core/testing

### Is this a regression?

Not sure, this is my first time to use Ivy.
The error is occurred in Ivy mode, but not in normal mode.

### Description

When running unit test using overrideProvider(), compileComponents(), and createComponent(), there is an error occurred: 

> moduleDef.imports is not iterable.

This error requires three conditions to occur: 
1. Iyv mode.
2. Import a module containing a function as imports, such as HttpClientTestingModule.
3. Using overrideProvider() and re-compile components.

#### Spec Code

```javascript
beforeEach(async(() => {
    TestBed.configureTestingModule({
      // Import the module containing a function in imports.
      imports: [HttpClientTestingModule],
      declarations: [
        AppComponent
      ],
      // Initialization for providers.
      providers: [{ provide: InjectedString, useValue: { value: 'Initial...' } }],
    }).compileComponents();
  }));

  it('overrideProvider with HttpClientTestingModule should work', async(() => {
    // Override provider with value 'Success!'.
    TestBed.overrideProvider(InjectedString, { useValue: { value: 'Success!' } }).compileComponents();

    const fixture = TestBed.createComponent(AppComponent);
    fixture.detectChanges();
    const compiled = fixture.debugElement.nativeElement;
    expect(compiled.querySelector('.injected').textContent).toContain('Success!');
  }));
```

#### tsconfig.spec.json
 ```javascript
 {
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "./out-tsc/spec",
    "types": [
      "jasmine",
      "node"
    ]
  },
  "files": [
    "src/test.ts",
    "src/polyfills.ts"
  ],
  "include": [
    "src/**/*.spec.ts",
    "src/**/*.d.ts"
  ],
  "angularCompilerOptions": {
    // If disable Ivy, the testing of overrideProvider() will pass.
    "enableIvy": true
  }
}
 ```

## 🔬 Minimal Reproduction

https://github.com/zalic0814/ivy-override-provider

### Error
<pre><code>
Failed: moduleDef.imports is not iterable

at <Jasmine>
at R3TestBedCompiler.applyProviderOverridesToModule (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:2104:1)
at R3TestBedCompiler.applyProviderOverridesToModule (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:2105:1)
at R3TestBedCompiler.compileTestModule (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:2386:1)
at R3TestBedCompiler.finalize (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:1874:1)
at TestBedRender3.get testModuleRef [as testModuleRef] (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:3249:1)
at TestBedRender3.inject (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:3110:1)
at TestBedRender3.createComponent (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:3200:1)
at Function.createComponent (http://localhost:9876/_karma_webpack_/node_modules/@angular/core/__ivy_ngcc__/fesm2015/testing.js:3013:1)
at UserContext.<anonymous> (http://localhost:9876/_karma_webpack_/src/app/app.component.spec.ts:22:29)
at ZoneDelegate.invoke (http://localhost:9876/_karma_webpack_/node_modules/zone.js/dist/zone-evergreen.js:365:1)
</code></pre>

## 🌍  Your Environment

Image 9

### PR DESCRIPTION
**Pull request by crisbeto**
Fixes an error that is thrown when a provider is overridden in `TestBed`, if the module definition of one of the imported modules uses a function for the `imports` that is set via `setNgModuleScope`. The problem was that we have a `for...of` loop that assumes that the imports are an array, but they can also be a function. This was handled correctly in other places, but the one here was missed.

Note that the above-mentioned error is only thrown at runtime when the code is transpiled to es6. In es5 TS generates a call to a helper that handles the error silently so the attached unit test only fails in es6.

Fixes #34623.

### CODE REVIEW

File: packages/core/test/test_bed_spec.ts [hunk 1]
```diff
@@ -1002,4 +1003,49 @@ describe('TestBed', () => {
...
+                .configureTestingModule({
+                  imports: [TestingModule],
+                  declarations: [AppComponent],
+                  providers: [{provide: InjectedString, useValue: {value: 'initial'}}],
+                })
+                .compileComponents();
+
+            let fixture: ComponentFixture<AppComponent>;
+
+            expect(() => {
```
Comment (AndrewKushnir) [inline packages/core/test/test_bed_spec.ts]: @crisbeto I'd propose to update test description a bit to say `should handle provider overrides when module imports are provided as a function` and remove the `expect(() => { ... }).not.toThrow()` wrapper, i.e. keep the test more generic.

**Review by AndrewKushnir — APPROVED** [review]
LGTM (with one minor comment), thanks for the fix @crisbeto 👍

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:288007601:BASE:288007631:1578074375221:38c23646)

File: packages/core/test/test_bed_spec.ts (hunk 1 above)
Comment (crisbeto) [reply packages/core/test/test_bed_spec.ts]: Done.

**Pull request closed without merging**